File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
            nums.sort()
            ans = []
            for i, val in enumerate(nums):
                if i > 0 and nums[i] == nums[i-1]:
                    continue

                l, r = i+1 , len(nums)-1
                while l < r:
                    target = val + nums[l] + nums[r]
                    if target > 0:
                        r -= 1
                    elif target < 0:
                        l += 1
                    else:
                        ans.append([val , nums[l] , nums[r]])
                        l += 1
                        while l < r and nums[l] == nums[l-1]:
                            l += 1

            return ans

File: 15-3sum/15-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        	vector<vector<int>> v ;

	sort(nums.begin() , nums.end());
	int n = nums.size();
	
	for (int i = 0 ; i < n ; i ++ ){
	if (i > 0 && nums[i] == nums[i-1]) continue;	

		int l = i+1 , r = n-1;
		while(l < r ){
			int sum = nums[i]+nums[l]+nums[r] ;
			if (sum == 0){
				v.push_back({nums[i] , nums[l] , nums[r]});
				l ++ ;
				while (nums[l] == nums[l-1] && l < r){
					l ++;
				}
			}else if (sum > 0){
				r --;
			}else {
				l ++;
			}
		}
		
	}


	return v ;
    }
};